File: mobile/android/core/commerce/src/main/kotlin/com/us/android/core/commerce/model/Paise.kt
package com.us.android.core.commerce.model

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

/**
 * An amount in Indian paise. 100 paise = ₹1.
 *
 * LB-A1 / LB-19. The backend carried money as `float64` rupees and crossed
 * the payments-service boundary as `int64(round(x * 100))`, so every hop was
 * a lossy IEEE-754 round trip. That is fixed server-side; this type is what
 * stops the app reintroducing it at the other end.
 *
 * A `@JvmInline value class` over `Long` costs nothing at runtime — it
 * compiles to a bare long — while making `Double` money a compile error
 * rather than a code-review question. The CI gate in
 * `.github/workflows/integration-commerce.yml` rejects a `Double` on any
 * money-named property in these modules, but the type system catches it
 * first and catches it everywhere.
 *
 * Division is deliberately absent. Splitting money is never a plain divide —
 * it has to allocate a remainder — and that allocation lives server-side in
 * `internal/tax`, which owns the rule and is property-tested against it. The
 * app displays what it is told.
 */
@JvmInline
@Serializable(with = PaiseSerializer::class)
value class Paise(val value: Long) : Comparable<Paise> {

    operator fun plus(other: Paise) = Paise(value + other.value)
    operator fun minus(other: Paise) = Paise(value - other.value)

    /**
     * Scale by a quantity. Quantity is the only multiplier money is
     * legitimately scaled by here; a percentage would need a rounding rule,
     * and that rule is the server's.
     */
    operator fun times(quantity: Int) = Paise(value * quantity)

    override fun compareTo(other: Paise): Int = value.compareTo(other.value)

    val isZero: Boolean get() = value == 0L
    val isPositive: Boolean get() = value > 0L

    /**
     * Renders "1,234.56" with Indian digit grouping (2,2,3 from the right:
     * ₹12,34,567.89, not ₹1,234,567.89).
     *
     * Integer arithmetic throughout — formatting money by dividing into a
     * Double is how a rounding bug reaches a customer's screen.
     */
    @Suppress("MagicNumber")
    fun format(): String {
        val negative = value < 0
        val abs = if (negative) -value else value
        val rupees = abs / 100
        val paise = abs % 100

        val digits = rupees.toString()
        val grouped = if (digits.length <= 3) {
            digits
        } else {
            val last3 = digits.takeLast(3)
            val rest = digits.dropLast(3)
            val chunks = mutableListOf<String>()
            var remaining = rest
            while (remaining.length > 2) {
                chunks.add(0, remaining.takeLast(2))
                remaining = remaining.dropLast(2)
            }
            if (remaining.isNotEmpty()) chunks.add(0, remaining)
            (chunks + last3).joinToString(",")
        }
        val body = "$grouped.${paise.toString().padStart(2, '0')}"
        return if (negative) "-$body" else body
    }

    /** "₹1,234.56", for display only. */
    fun formatWithSymbol(): String = "₹${format()}"

    companion object {
        val ZERO = Paise(0)

        /** 100 paise to the rupee. The one conversion factor in this type. */
        const val PAISE_PER_RUPEE = 100L

        /** For constants and config only, never for a user-entered amount. */
        fun ofRupees(rupees: Long) = Paise(rupees * PAISE_PER_RUPEE)
    }
}

/**
 * Wire format is a bare integer, always.
 *
 * Decoding REFUSES a fractional number. A `1234.56` in a paise field means
 * the sender is still thinking in rupees, and silently truncating it is
 * precisely how the original defect propagated between services. Better a
 * loud parse failure than a customer charged the wrong amount.
 */
object PaiseSerializer : KSerializer<Paise> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("Paise", PrimitiveKind.LONG)

    override fun serialize(encoder: Encoder, value: Paise) = encoder.encodeLong(value.value)

    override fun deserialize(decoder: Decoder): Paise = Paise(decoder.decodeLong())
}

File: Architecture/services/commerce-service/internal/money/paise.go
// Package money is the single representation of currency in commerce-service.
//
// LB-19. Before this package, commerce carried money as float64 rupees and
// crossed the payments-service boundary via int64(math.Round(x*100)). Every
// hop was a lossy IEEE-754 round trip, and the two services disagreed about
// what a rupee was: payments-service has been paise-native since its
// migration 006, commerce was not.
//
// Paise is a distinct type, not an alias, so an int64 that means "quantity"
// or "seconds" cannot be assigned to a money field by accident. Division is
// deliberately absent: splitting money is never a plain divide (it must
// allocate a remainder), so that operation lives in internal/tax where the
// allocation rule is explicit and tested.
package money

import (
	"database/sql/driver"
	"fmt"
	"strconv"
	"strings"
)

// Paise is an amount in Indian paise. 100 paise = ₹1.
//
// A negative Paise is legal — it represents a debit direction in the
// inventory/refund ledgers — so callers that require a non-negative amount
// must assert it themselves. Use MustNonNegative at trust boundaries.
type Paise int64

// Zero is the additive identity, named so call sites read as intent rather
// than as a magic literal.
const Zero Paise = 0

// FromRupees converts whole rupees. It exists for constants and config
// (a ₹499 free-shipping threshold), never for arithmetic on a user amount.
func FromRupees(r int64) Paise { return Paise(r * 100) }

// Add returns a + b.
func (p Paise) Add(o Paise) Paise { return p + o }

// Sub returns a - b.
func (p Paise) Sub(o Paise) Paise { return p - o }

// MulQty scales an amount by an integer quantity. Quantity is the only
// multiplier money is ever legitimately scaled by in this domain — a
// percentage is applied through internal/tax, which owns the rounding rule.
func (p Paise) MulQty(qty int) Paise { return p * Paise(qty) }

// IsZero reports whether the amount is exactly zero.
func (p Paise) IsZero() bool { return p == 0 }

// IsNegative reports whether the amount is below zero.
func (p Paise) IsNegative() bool { return p < 0 }

// Int64 returns the raw paise count, for SQL parameters and provider APIs
// that take an integer minor unit (Razorpay's `amount`, for one).
func (p Paise) Int64() int64 { return int64(p) }

// String renders "₹1,234.56"-style text for logs and human-facing surfaces.
// It is never used to compute anything.
func (p Paise) String() string {
	neg := p < 0
	v := int64(p)
	if neg {
		v = -v
	}
	major, minor := v/100, v%100
	s := fmt.Sprintf("%d.%02d", major, minor)
	if neg {
		s = "-" + s
	}
	return s
}

// MarshalJSON emits a bare integer. The wire contract for every money field
// is minor units; a JSON number with a decimal point in a money field is a
// bug that this makes impossible to introduce silently.
func (p Paise) MarshalJSON() ([]byte, error) {
	return []byte(strconv.FormatInt(int64(p), 10)), nil
}

// UnmarshalJSON accepts an integer or an integer-valued string. It REJECTS a
// fractional number: "1234.56" in a paise field means the caller is still
// thinking in rupees, and silently truncating that is how the original
// float defect propagated.
func (p *Paise) UnmarshalJSON(b []byte) error {
	s := strings.TrimSpace(strings.Trim(string(b), `"`))
	if s == "" || s == "null" {
		*p = 0
		return nil
	}
	if strings.ContainsAny(s, ".eE") {
		return fmt.Errorf("money: %q is not an integer minor amount; money must be sent in paise", s)
	}
	v, err := strconv.ParseInt(s, 10, 64)
	if err != nil {
		return fmt.Errorf("money: parse paise %q: %w", s, err)
	}
	*p = Paise(v)
	return nil
}

// Value implements driver.Valuer so Paise binds directly to a BIGINT column.
func (p Paise) Value() (driver.Value, error) { return int64(p), nil }

// Scan implements sql.Scanner for BIGINT columns.
//
// It deliberately accepts float64 as well, because during the dual-write
// migration window a value may still arrive from a NUMERIC(12,2) mirror.
// That path rounds half-away-from-zero exactly once, at the boundary, and is
// removed with the mirrors in the contraction phase.
func (p *Paise) Scan(src any) error {
	switch v := src.(type) {
	case nil:
		*p = 0
		return nil
	case int64:
		*p = Paise(v)
		return nil
	case int32:
		*p = Paise(v)
		return nil
	case float64:
		// Legacy NUMERIC rupees mirror. Round half away from zero.
		if v < 0 {
			*p = Paise(int64(v*100 - 0.5))
		} else {
			*p = Paise(int64(v*100 + 0.5))
		}
		return nil
	case []byte:
		return p.UnmarshalJSON(v)
	case string:
		return p.UnmarshalJSON([]byte(v))
	default:
		return fmt.Errorf("money: cannot scan %T into Paise", src)
	}
}

// MustNonNegative returns an error when the amount is negative. Trust
// boundaries (request bodies, provider responses, event payloads) call this
// before the value reaches the domain.
func MustNonNegative(field string, p Paise) error {
	if p < 0 {
		return fmt.Errorf("money: %s must not be negative (got %s)", field, p)
	}
	return nil
}

// Sum adds a slice, used for line totals.
func Sum(ps ...Paise) Paise {
	var t Paise
	for _, p := range ps {
		t += p
	}
	return t
}
